### PR TITLE
feat: SPA and `index.html` serving

### PR DESCRIPTION
**Pull request by Antosser**
Copy/pasted every important bit from the previous PR.

Everything looks great except for one small problem: When the `--spa` flag is used, the `--index` flag has to be enabled as well. But the code is really split up into a lot of files and I don't know where I can set `cli.index` to `true` if `cli.spa` is `true` in the code

### CODE REVIEW

File: src/addon/file_server/mod.rs [hunk 1]
```diff
@@ -128,33 +133,79 @@ impl<'a> FileServer {
...
+                                File {
+                                    path: filepath,
+                                    metadata: file.metadata().await?,
+                                    file,
+                                },
+                                CacheControlDirective::MaxAge(2500),
+                            )
+                            .await;
+                        }
+                    }
```
Comment (LeoBorai) [inline src/addon/file_server/mod.rs]: I think the `'dir` tag is not required in this context. Can we remove it?

File: src/addon/file_server/mod.rs [hunk 2]
```diff
@@ -128,33 +133,79 @@ impl<'a> FileServer {
...
-                    .body(Body::from(err.to_string()))
-                    .expect("Failed to build response")),
-                _ => Ok(HttpResponseBuilder::new()
-                    .status(StatusCode::BAD_REQUEST)
-                    .body(Body::from(err.to_string()))
-                    .expect("Failed to build response")),
-            },
+            Err(err) => {
+                if self.config.spa() {
+                    make_http_file_response(
```
Comment (LeoBorai) [inline src/addon/file_server/mod.rs]: Short circuit here please, by returning instead of `else`.

File: src/server/handler/mod.rs [hunk 3]
```diff
@@ -55,7 +55,7 @@ impl From<Arc<Config>> for HttpHandler {
             };
         }
 
-        let file_server = FileServer::new(config.root_dir());
+        let file_server = FileServer::new(config.root_dir(), config.clone());
```
Comment (LeoBorai) [inline src/server/handler/mod.rs]: Why not just borrowing `config`?

```suggestion
        let file_server = FileServer::new(&config);
```

**Review by LeoBorai — COMMENTED** [review]
Looking better!

**Review by LeoBorai — APPROVED** [review]

**Comment by LeoBorai** [comment]
> Copy/pasted every important bit from the previous PR.
> 
> 
> 
> Everything looks great except for one small problem: When the `--spa` flag is used, the `--index` flag has to be enabled as well. But the code is really split up into a lot of files and I don't know where I can set `cli.index` to `true` if `cli.spa` is `true` in the code

IIRC Clap allows you to have dependencies between arguments.

I think we can merge this and have that improvement as iteration.

**Comment by Antosser** [comment]
The problem from the description is still there. In the current state `--spa` will work weirdly without `--index`

**Pull request merged**